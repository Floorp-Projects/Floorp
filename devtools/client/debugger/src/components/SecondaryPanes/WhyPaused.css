/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.why-paused {
  display: flex;
  flex-direction: column;
  justify-content: center;
  border-bottom: 1px solid var(--theme-splitter-color);
  background-color: hsl(54, 100%, 92%);
  color: var(--theme-body-color);
  font-size: 12px;
  cursor: default;
  min-height: 44px;
  padding: 6px;
  white-space: normal;
  font-weight: bold;
}

.why-paused > div {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.why-paused .info.icon {
  align-self: center;
  padding-right: 4px;
  margin-inline-start: 14px;
  margin-inline-end: 3px;
}

.why-paused .pause.reason {
  display: flex;
  flex-direction: column;
  padding-right: 4px;
}

.theme-dark .secondary-panes .why-paused {
  background-color: hsl(42, 37%, 19%);
  color: hsl(43, 94%, 81%);
}

.why-paused .message {
  font-style: italic;
  font-weight: 100;
}

.why-paused .mutationNode {
  font-weight: normal;
}

.why-paused .message.warning {
  color: var(--theme-graphs-full-red);
  font-family: var(--monospace-font-family);
  font-size: 10px;
  font-style: normal;
}
