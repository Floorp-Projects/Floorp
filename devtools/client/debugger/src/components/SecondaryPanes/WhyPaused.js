/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React, { PureComponent } from "react";
import { connect } from "../../utils/connect";
import AccessibleImage from "../shared/AccessibleImage";
import actions from "../../actions";

import Reps from "devtools/client/shared/components/reps/index";
const {
  REPS: { Rep },
  MODE,
} = Reps;

import { getPauseReason } from "../../utils/pause";
import {
  getCurrentThread,
  getPaneCollapse,
  getPauseReason as getWhy,
} from "../../selectors";

import "./WhyPaused.css";

class WhyPaused extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { hideWhyPaused: "" };
  }

  componentDidUpdate() {
    const { delay } = this.props;

    if (delay) {
      setTimeout(() => {
        this.setState({ hideWhyPaused: "" });
      }, delay);
    } else {
      this.setState({ hideWhyPaused: "pane why-paused" });
    }
  }

  renderExceptionSummary(exception) {
    if (typeof exception === "string") {
      return exception;
    }

    const { preview } = exception;
    if (!preview || !preview.name || !preview.message) {
      return;
    }

    return `${preview.name}: ${preview.message}`;
  }

  renderMessage(why) {
    const { type, exception, message } = why;

    if (type == "exception" && exception) {
      // Our types for 'Why' are too general because 'type' can be 'string'.
      // $FlowFixMe - We should have a proper discriminating union of reasons.
      const summary = this.renderExceptionSummary(exception);
      return <div className="message warning">{summary}</div>;
    }

    if (type === "mutationBreakpoint" && why.nodeGrip) {
      const { nodeGrip, ancestorGrip, action } = why;
      const {
        openElementInInspector,
        highlightDomElement,
        unHighlightDomElement,
      } = this.props;

      const targetRep = Rep({
        object: nodeGrip,
        mode: MODE.TINY,
        onDOMNodeClick: () => openElementInInspector(nodeGrip),
        onInspectIconClick: () => openElementInInspector(nodeGrip),
        onDOMNodeMouseOver: () => highlightDomElement(nodeGrip),
        onDOMNodeMouseOut: () => unHighlightDomElement(),
      });

      const ancestorRep = ancestorGrip
        ? Rep({
            object: ancestorGrip,
            mode: MODE.TINY,
            onDOMNodeClick: () => openElementInInspector(ancestorGrip),
            onInspectIconClick: () => openElementInInspector(ancestorGrip),
            onDOMNodeMouseOver: () => highlightDomElement(ancestorGrip),
            onDOMNodeMouseOut: () => unHighlightDomElement(),
          })
        : null;

      return (
        <div>
          <div className="message">{why.message}</div>
          <div className="mutationNode">
            {ancestorRep}
            {ancestorGrip ? (
              <span className="why-paused-ancestor">
                {action === "remove"
                  ? L10N.getStr("whyPaused.mutationBreakpointRemoved")
                  : L10N.getStr("whyPaused.mutationBreakpointAdded")}
                {targetRep}
              </span>
            ) : (
              targetRep
            )}
          </div>
        </div>
      );
    }

    if (typeof message == "string") {
      return <div className="message">{message}</div>;
    }

    return null;
  }

  render() {
    const { endPanelCollapsed, why } = this.props;
    const reason = getPauseReason(why);

    if (!why || !reason || endPanelCollapsed) {
      return <div className={this.state.hideWhyPaused} />;
    }

    return (
      <div className="pane why-paused">
        <div>
          <div className="info icon">
            <AccessibleImage className="info" />
          </div>
          <div className="pause reason">
            {L10N.getStr(reason)}
            {this.renderMessage(why)}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  endPanelCollapsed: getPaneCollapse(state, "end"),
  why: getWhy(state, getCurrentThread(state)),
});

export default connect(mapStateToProps, {
  openElementInInspector: actions.openElementInInspectorCommand,
  highlightDomElement: actions.highlightDomElement,
  unHighlightDomElement: actions.unHighlightDomElement,
})(WhyPaused);
