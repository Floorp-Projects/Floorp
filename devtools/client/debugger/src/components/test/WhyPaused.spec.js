/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";
import WhyPaused from "../SecondaryPanes/WhyPaused.js";

function render(why, delay) {
  const props = { why, delay };

  const component = shallow(<WhyPaused.WrappedComponent {...props} />);

  return { component, props };
}

describe("WhyPaused", () => {
  it("should pause reason with message", () => {
    const why = {
      type: "breakpoint",
      message: "bla is hit",
    };
    const { component } = render(why);
    expect(component).toMatchSnapshot();
  });

  it("should show pause reason with exception details", () => {
    const why = {
      type: "exception",
      exception: {
        class: "ReferenceError",
        isError: true,
        preview: {
          name: "ReferenceError",
          message: "o is not defined",
        },
      },
    };

    const { component } = render(why);
    expect(component).toMatchSnapshot();
  });

  it("should show pause reason with exception string", () => {
    const why = {
      type: "exception",
      exception: "Not Available",
    };

    const { component } = render(why);
    expect(component).toMatchSnapshot();
  });

  it("should show an empty div when there is no pause reason", () => {
    const why = undefined;

    const { component } = render(why);
    expect(component).toMatchSnapshot();
  });
});
