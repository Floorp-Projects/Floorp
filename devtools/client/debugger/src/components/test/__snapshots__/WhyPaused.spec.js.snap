// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhyPaused should pause reason with message 1`] = `
<div
  className="pane why-paused"
>
  <div>
    <div
      className="info icon"
    >
      <AccessibleImage
        className="info"
      />
    </div>
    <div
      className="pause reason"
    >
      Paused on breakpoint
      <div
        className="message"
      >
        bla is hit
      </div>
    </div>
  </div>
</div>
`;

exports[`WhyPaused should show an empty div when there is no pause reason 1`] = `
<div
  className=""
/>
`;

exports[`WhyPaused should show pause reason with exception details 1`] = `
<div
  className="pane why-paused"
>
  <div>
    <div
      className="info icon"
    >
      <AccessibleImage
        className="info"
      />
    </div>
    <div
      className="pause reason"
    >
      Paused on exception
      <div
        className="message warning"
      >
        ReferenceError: o is not defined
      </div>
    </div>
  </div>
</div>
`;

exports[`WhyPaused should show pause reason with exception string 1`] = `
<div
  className="pane why-paused"
>
  <div>
    <div
      className="info icon"
    >
      <AccessibleImage
        className="info"
      />
    </div>
    <div
      className="pause reason"
    >
      Paused on exception
      <div
        className="message warning"
      >
        Not Available
      </div>
    </div>
  </div>
</div>
`;
